### ISSUE TITLE
Getting all rows to be exported to excel using resultSet

### ISSUE THREAD

**Issue by a2441918** [oracle/node-oracledb#786]
Hi, I am using resultSet class to get data from my DB and populate data on a table. I also have an 'Export' button which exports all the data which is retrieved to an excel file. It doesnt seem to work when I am using ResultSet. Please find my code below for your reference.

    router.post('/report/', jsonParser, function (req, res) {
    var data = req.body,
        name= data.name,
        startRow = data.startRow,
        numRows = data.numRows,
        sortCol = assert.simpleSqlName(data.sortCol),
        sortDir = assert.validSortOrder(data.sortDir);

    var countQuery = 'SELECT COUNT(*) ' +
        'FROM this_view ' +
        'WHERE Status = \'Present\' ';

    var query = 'SELECT t.NAME "NAME", ' +
        't.ADDRESS"ADDRESS", ' +
        'FROM this_view t ' +
        'WHERE Status = \'Present\' ';

    var seg,
        orderBy,
        offset;

    if(name !== null && name !== ""){
        seg = ' AND "NAME" LIKE \'%'+ name + '%\'';
        query += seg;
        countQuery += seg;
    }

    orderBy = ' ORDER BY UPPER(' + sortCol + ') ' + sortDir;
    offset = ' OFFSET :start_row ROWS FETCH NEXT :num_rows ROWS ONLY';

    query += orderBy;
    query += offset;

    logger.info("Begin: " + (new Date().toString()));

    async.parallel({
        rows: function (callback) {
            pool.getConnection(function (err, connection) {
                logger.info("Begin Connection: " + (new Date().toString()));
                if (err) {
                    logger.error(err.message);
                    return;
                }

                logger.info("Begin execute: " + (new Date().toString()));

                connection.execute(
                    query,
                    {
                        start_row: startRow,
                        num_rows: numRows
                    },
                    {
                        resultSet: true,
                        prefetchRows: 1000
                    },
                    function (err, result) {
                        logger.info("End execute: " + (new Date().toString()));
                        if (err) {
                            logger.error(err.message);
                            callback("Something broke in the first thing");
                            doRelease(connection);
                            return;
                        }
                        var procJson = [];

                        function fetchRowsFromRS(connection, resultSet, numRows) {
                            resultSet.getRows(
                                numRows,  // get this many rows
                                function (err, rows) {
                                    if (err) {
                                        console.error(err);
                                        doClose(connection, resultSet); // always close the result set
                                    } else if (rows.length > 0) {

                                        /**
                                         *  For each row in the result, pushes a new object to the rows array
                                         *  In each new object, the key is assigned and the result row value set
                                         */
                                        for (var i = 0; i < rows.length; i++) {
                                            procJson.push({});
                                            for (var j = 0; j < resultSet.metaData.length; j++) {
                                                procJson[i][resultSet.metaData[j].name.toLowerCase()] = rows[i][j];
                                            }
                                        }

                                        if (rows.length <= numRows) // might be more rows
                                            fetchRowsFromRS(connection, resultSet, numRows);
                                        else
                                            doClose(connection, resultSet); // always close the result set
                                    } else {
                                        logger.info("Send JSON: " + (new Date().toString()));
                                        callback(null, procJson);
                                        logger.info("JSON Sent: " + (new Date().toString()));
                                        doClose(connection, resultSet); // always close the result set
                                    }
                                });
                        }
                        fetchRowsFromRS(connection, result.resultSet, numRows);
                    });
            });
        },
        totalRows: function (callback) {
            pool.getConnection(function (err, connection) {
                logger.info("Begin Connection: " + (new Date().toString()));
                if (err) {
                    logger.error(err.message);
                    return;
                }

                logger.info("Begin execute: " + (new Date().toString()));

                connection.execute(
                    countQuery,
                    function (err, result) {
                        logger.info("End execute: " + (new Date().toString()));
                        if (err) {
                            logger.error(err.message);
                            callback("Something broke");
                            doRelease(connection);
                            return;
                        }

                        logger.info("Send JSON: " + (new Date().toString()));
                        console.log(result.rows);
                        callback(null, result.rows[0][0]);
                        logger.info("JSON Sent: " + (new Date().toString()));

                        doRelease(connection);
                    });
            });
        }
    }, function(err, result){
        if(err){
            logger.error(err);
            res.status(404);
            res.send('No Entries');
        }
        else {
            res.send(result);
        }
    });
    });

**Comment by a2441918** [comment]
@cjbj Any suggestions?

**Comment by dmcghan** [comment]
@a2441918 Please don't post a question and then ask for suggestions right after. This isn't paid support. 

You need to learn some basic troubleshooting skills. Try setting up small test cases to see how things behave. Break things down and then build them back up.

Here's a small test that should help you understand the issue.

```javascript
const oracledb = require('oracledb');
const config = require('./dbConfig.js'); 
let conn;

async function testUndefined() {
  try {
    conn = await oracledb.getConnection(config);

    const result = await conn.execute(
      `select 1 from dual 
      offset :start_row rows fetch next :num_rows rows only`,
      {
        start_row: undefined,
        num_rows: undefined
      }
    );

    console.log(result.rows);
  } catch (err) {
    console.error(err);
  } finally {
    if (conn) {
      try {
        await conn.close();
      } catch (err) {
        console.error(err);
      }
    }
  }
}

testUndefined();
```

**Comment by a2441918** [comment]
@dmcghan @cjbj Sorry about that. I was not expecting an answer soon. I think I should use queryStream API with resultSet to get the whole data. I am in the learning process of understanding this module and that is why I post so many questions. Sorry about that again.

**Comment by dmcghan** [comment]
@a2441918 You may prefer the QueryStream API, but it doesn't change the issue. The QueryStream API uses a ResultSet under the hood. You will not get rows until you account for the fact that when you want all rows `start_row` and `num_rows` are `undefined`. You should not be appending the OFFSET or FETCH clauses in that case.

**Comment by a2441918** [comment]
Sorry, if I am reiterating this. If I dont pass in an offset, resultSet will fetch all rows since start_row and now_rows will be undefined/not passed in?

**Comment by a2441918** [comment]
Sorry, if I am reiterating this. If I dont pass in an offset, resultSet will fetch all rows since start_row and now_rows will be undefined/not passed in?

**Comment by sagiegurari** [comment]
if you just want to pull all rows and you have enough memory to hold it all in memory, you can use simple-oracledb query function.

//do this only once in the entire app lifetime
//see [init simple oracledb](https://github.com/sagiegurari/simple-oracledb#usage)
var oracledb = require('oracledb');
var SimpleOracleDB = require('simple-oracledb');

//create a pool as shown in oracledb docs
var pool = ....

//once a request comes in, use the pool
//see [pool.run](https://github.com/sagiegurari/simple-oracledb#Pool+run)
pool.run((connection, callback) => {
  //run some query and the output will be available in the 'run' callback
  //see [connection.query](https://github.com/sagiegurari/simple-oracledb#Connection+query)
  connection.query('SELECT FIELD1, FIELD2 FROM MY_TABLE WHERE ID = :0', [110], callback);
}, (error, result) => {
  //do something with the result/error like generating excel and pushing to the http response
  //by default connection.query will pull ALL results so make sure you have enough memory for that
});

**Comment by sagiegurari** [comment]
11k rows and 100+ cols is definitely too much, especially when you have concurrent user requests.
So ya, like @dmcghan stated, querystream would be best option for that.

**Comment by a2441918** [comment]
@sagiegurari Yeah. Trying to make it work by removing OFFSET from my query and removing the bind variables under connection.execute but in vain. Trying other approaches too until dmcghan corrects my understanding.

**Comment by dmcghan** [comment]
> If I dont pass in an offset, resultSet will fetch all rows since start_row and now_rows will be undefined/not passed in?

You tell me.

**Comment by a2441918** [comment]
Looks like if its undefined, it should give me all rows.

**Comment by a2441918** [comment]
But I guess the payload is too much for the client to handle and I guess I have to make an excel in the server side and pass that to the client..

**Comment by dmcghan** [comment]
> Looks like if its undefined, it should give me all rows.

Does it really? Show me a test.

**Comment by a2441918** [comment]
function resultSetTest() {
        return oracledb.getConnection(poolAttrs, function(err, connection) {
            if (err) {
                logger.error(err.message);
                return;
            }
            connection.execute(
                'SELECT * FROM this_view offset :start_row rows fetch next :num_rows rows only',
                {
                    start_row: undefined,
                    num_rows: undefined
                },
                function (err, result) {
                    if (err) {
                        logger.error(err.message);
                        callback("Something broke in the first thing");
                        doRelease(connection);
                        return;
                    }

                    console.log(result);
                    doRelease(connection);
                }
            )

        });
    }

    resultSetTest();

This is my test. rows is undefined. This is the log of result which I did.

        { rows: [],
      resultSet: undefined,
      outBinds: undefined,
      rowsAffected: undefined,
      metaData:
       [ { name: 'NAME' },
         { name: 'ADDRESS' },
         { name: 'CITY' },
         { name: 'COUNTRY' },
         { name: 'ID' },
         { name: 'PHONE' },
         { name: 'MOBILE' },
         { name: 'USERNAME' } ] }

**Comment by cjbj** [comment]
This is what SQL*Plus gives in a couple of cases:

```
SQL> SELECT * FROM locations offset null rows fetch next null rows only;

no rows selected

SQL> SELECT * FROM locations offset 0 rows fetch next 0 rows only;

no rows selected
```

Seems you have to pass in 'valid' values....

**Comment by a2441918** [comment]
Yeah, @cjbj. You can use 
    SELECT * FROM locations offset 0 rows fetch next 10 rows only;

to get the first 10 rows. You can also change the value near offset to any number and will fetch between that range.

When I try to use a connection without resultSet, I get a memory allocation error. So trying to use resultSet but get the whole payload to export it to a excel file.

**Comment by dmcghan** [comment]
@a2441918 

> This is my test. rows is undefined. This is the log of result which I did.

`rows` is not `undefined`. `rows` is an empty array.

So your test seems to show that when you use the OFFSET and FETCH clauses with `undefined` values, you get no rows back. But you want to get the entire result set back. 

How could you modify the query to get all the rows back?

**Comment by a2441918** [comment]
@dmcghan I could pass in the totalRows which I got from executing "SELECT COUND(*) FROM this_view" as numRows. Now, numRows will be ~11000. I pass that into the query. 

So my query will look like this.
    SELECT * FROM this_view offset 0 rows fetch next 11000 rows only

**Comment by dmcghan** [comment]
@a2441918 I suppose you could do that. But isn't there a simpler solution? Couldn't you just omit the OFFSET and FETCH clauses?

**Comment by a2441918** [comment]
Yes @dmcghan but I get "NJS-024 Memory Allocation Failed" when I try to get the whole results when I use just "SELECT * FROM this_view" ...

**Comment by dmcghan** [comment]
@a2441918 Show me your test.

**Comment by a2441918** [comment]
This is my test.

             function resultSetTest() {
        return oracledb.getConnection(poolAttrs, function(err, connection) {
            if (err) {
                logger.error(err.message);
                return;
            }
            connection.execute(
                'SELECT * FROM this_view',
                function (err, result) {
                    if (err) {
                        logger.error("resultSetTestError", err.message);
                        doRelease(connection);
                        return;
                    }

                    console.log("resultSet", result);
                    doRelease(connection);
                }
            )

        });
    }

    resultSetTest();

This is the error which my logger gave.

> {"name":"Logger","pid":14192,"level":50,"msg":"resultSetTestError NJS-024: memory allocation failed","time":"2017-11-09T14:53:06.301Z","v":0}

**Comment by dmcghan** [comment]
@a2441918 The name is `resultSetTest`, but you don't use a ResultSet. Rewrite the test to use a ResultSet. You can choose to use the standard ResultSet API with recursion or use the QueryStream API.

**Comment by sagiegurari** [comment]
Even without resultset, wonder why it would fail.
By default I think the limit is only 100 rows, no?
So maybe there is another issue here? Maybe just limit to fewer rows to see if that's really lack of memory but still it looks strange. It's not a lot of data

**Comment by dmcghan** [comment]
@sagiegurari Fair point! Perhaps the view has 1000 clobs? ;)

@a2441918 Did you really run that test?

**Comment by a2441918** [comment]
function resultSetTest() {
        return oracledb.getConnection(poolAttrs, function(err, connection) {
            if (err) {
                logger.error(err.message);
                return;
            }
            connection.execute(
                'SELECT * FROM this_view',
                {},
                {
                    resultSet: true,
                    prefetchRows: 1000
                },
                function (err, result) {
                    if (err) {
                        logger.error("resultSetTestError", err.message);
                        doRelease(connection);
                        return;
                    }

                    function fetchRowsFromRS(connection, resultSet, numRows) {
                        resultSet.getRows(
                            numRows,  // get this many rows
                            function (err, rows) {
                                if (err) {
                                    console.error(err);
                                    doClose(connection, resultSet); // always close the result set
                                } else if (rows.length > 0) {

                                    console.log(rows.length);

                                    if (rows.length <= numRows) // might be more rows
                                        fetchRowsFromRS(connection, resultSet, numRows);
                                    else
                                        doClose(connection, resultSet); // always close the result set
                                } else {
                                    logger.info("Send JSON: " + (new Date().toString()));
                                    logger.info("JSON Sent: " + (new Date().toString()));
                                    doClose(connection, resultSet); // always close the result set
                                }
                            });
                    }

                    fetchRowsFromRS(connection, result.resultSet, 1000);

                }
            )

        });
    }

    resultSetTest();

So my console prints this.

    1000
    1000
    1000
    1000
    1000
    1000
    1000
    1000
    1000 
    1000
   998 

I'll try increasing numRows to 10000 and post the result.

@dmcghan Yeah, I did and thats how I got the logs

**Comment by a2441918** [comment]
@dmcghan @sagiegurari Bumping them upto 10000 works. I guess its an issue with the client trying to process that much information and convert to an excel file

**Comment by dmcghan** [comment]
@a2441918 Something isn't adding up here. You're saying when you use the default [`maxRows`](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#propexecmaxrows) of 100 you get a segfault, but if you use a `prefetchRows` of 1000 you don't.

I'm guessing you're setting [oracledb.maxRows](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#-328-oracledbmaxrows) much higher somewhere else. Can you review your code?

By the way, a better test is one file that includes everything - including where you require oracledb in. An even better test provides the DDL needed to recreate the objects your using or uses objects in a default schema like HR.

> So my console prints this.

So you're not getting errors and you're getting all the rows? Do you see how this will work for you?


> I'll try increasing numRows to 10000 and post the result.

That may be too high. Also, watch `prefetchRows` - that reduces network roundtrips by buffering data so don't set it too high either.

**Comment by a2441918** [comment]
Yes, maxRows is set to 400000.

**Comment by cjbj** [comment]
@a2441918 Here's the [doc on `maxRows`](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#propdbmaxrows).  I'd never set it bigger than few thousand, at most.

If Node.js itself can't cope with the memory required to hold all of your data, then you will have to stream that data somehow, either with a real Stream or by writing it in batches.

**Comment by dmcghan** [comment]
@a2441918 As Chris says, you'll need to stream it or batch it.

If the user is in an app looking at data in some kind of table/view with pagination that goes from one set of rows to the next, then you can batch it with `oracledb.execute`. In this case, you'd use the OFFSET and FETCH clauses and make sure that maxRows was above the FETCH value.

However, if you're downloading all the rows (or any quantity that exceeds that Node.js can handle at once), then you need a real Stream to get a read consistent image of the data. In that case, you would not use the OFFSET or FETCH clauses.

**Comment by cjbj** [comment]
There were some changes in v2 in this area.  See https://blogs.oracle.com/opal/node-oracledb-v2-query-methods-and-fetch-tuning and https://github.com/oracle/node-oracledb/blob/v2.0.15/CHANGELOG.md
